### ISSUE TITLE
Using RewriteHandler with AsyncContext.dispatch() and HttpServletRequestWrapper not possible

### ISSUE THREAD

**Issue by joakime** [eclipse/jetty.project#1642]
If you have an HttpServletRequestWrapper in use, and also have RewriteHandler rules setup, then you cannot use AsyncContext.dispatch() as it will trigger an exception in RewriteHandler.

```
06/22/17 18:39:05 [WARNING] /asyncdemo/foo/bar
java.lang.ClassCastException: org.eclipse.jetty.demo.AntiPatternRequestWrapper cannot be cast to org.eclipse.jetty.server.Request
        at org.eclipse.jetty.rewrite.handler.RuleContainer.apply(RuleContainer.java:188)
        at org.eclipse.jetty.rewrite.handler.RuleContainer.matchAndApply(RuleContainer.java:149)
        at org.eclipse.jetty.rewrite.handler.RewriteHandler.handle(RewriteHandler.java:330)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
        at org.eclipse.jetty.server.Server.handleAsync(Server.java:609)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:334)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:110)
        at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
        at org.eclipse.jetty.util.thread.Invocable.invokePreferred(Invocable.java:128)
        at org.eclipse.jetty.util.thread.Invocable$InvocableExecutor.invoke(Invocable.java:222)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:294)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:199)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:673)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:591)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by joakime** [comment]
@gregw what are your thoughts on a good jetty subproject/module for a unit test for this?
I'm thinking jetty-webapp, but that will cause jetty-rewrite to be brought in as a test-dependency.
Wonder if somewhere in /tests/ is appropriate.

**Comment by joakime** [comment]
A working fix has been commited to branch `issue-1642` in commit 95d4390.
Still needs a good unit test though.

**Comment by joakime** [comment]
merged into jetty-9.4.x
